Dynamo: support proxying tensor subclass constructors, including with non-fx types

bdhirsh opened the pull request:
For a while, we haven't really had good support for tensor subclass constructors going **inside** of the compiled region.

Two main example of that we've sort of half-handed in the last few months are:

(1) DTensor.to_local(), which is a wrapper around the constructor, that we explicitly allow_in_graph (cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire)
(2) Float8Tensor.to_float8(), which is similar, and also allowed_in_graph explicitly (cc @drisspg / @vkuzo)

This PR attempts to allow **all** traceable wrapper subclass constructors to go directly in the graph, when dynamo has a chance to intercept `SubclassCtr(*args)`.

To handle the DTensor case (where `Placement` and `DeviceMesh` cannot become nodes in the graph, but they are constructor inputs), I tried to add some infra on `VariableTracker` that tells us if a VT "supports going in the FX graph". For any inputs to the constructor that are not FX types, I close over them in the node's implementation, and convert them to their python constants (so AOT can implement them later).

cc @anijain2305, a review on the dynamo bits would be much appreciated.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123349
* __->__ #123350
* #124401
* #124400
* #124399
* #124398



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

bdhirsh committed:
Dynamo: support proxying tensor subclass constructors, including with non-fx types

[ghstack-poisoned]

bdhirsh committed:
Update on "Dynamo: support proxying tensor subclass constructors, including with non-fx types"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/lists.py:
```diff
@@ -63,6 +63,9 @@ def __init__(
     def _as_proxy(self):
         return [x.as_proxy() for x in self.items]
 
+    def can_be_proxied_in_fx(self):
+        return all(x.can_be_proxied_in_fx() for x in self.items)
```
Is this "can this Variable show up in the FX graph"? If so, lists should not show up in the FX IR

zou3519 replied:
Maybe I'm misunderstanding what `can_be_proxied_in_fx` means

bdhirsh replied:
Good catch - I forgot we support tuple and not list. and "can this Variable show up in the FX graph" is correct.

I didn't carefully audit every VariableTracker when adding this function, I basically did the minimal amount necessary to get a DTensor constructor test passing. But maybe I should if we want to actually land this PR (it is no longer needed to unblock anything, minus making Float8/DTensor constructor handling a bit cleaner).

ezyang commented on test/inductor/test_distributed_patterns.py:
```diff
@@ -63,13 +63,13 @@ def bw_post_hook(mod, gI, gO):
     m.register_forward_pre_hook(fw_pre_hook)
     m.register_forward_hook(fw_post_hook)
-    return m, torch.rand(2, 20, requires_grad=True)
+    return m, torch.rand(2, 20, requires_grad=True, device="cuda")
```
What's going on here?

bdhirsh replied:
oh welp, accidental commit - my env was broken in a way that broken inductor cpu compilation so I was testing with cuda (these tests were also for my ppFSDP testing, not relevant to this PR stack).

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -2651,3 +2651,69 @@ def wrapper(*args):
+# - Subclass()
+# - Subclass.__new__()
+# - Subclass.__init__()
+def proxy_fn_with_non_fx_args(fn, tx, args, kwargs):
```
@wanchaol does this subsume the thing you did for DTensor?

wanchaol replied:
yep, this looks bascially the lamaba closure function I did for DTensor

ezyang commented:
What is the difference between "can proxy in graph" vs "has an implementation of python_constant"?

bdhirsh commented:
A few things about this PR that make me unsure if it's worth landing.

(1) I don't strictly need this PR to unblock https://github.com/pytorch/torchtrain/issues/61 / https://github.com/pytorch/pytorch/issues/122459 (_dynamo.disable on __new__ is enough)

(2) @zou3519 convinced me that one alternative to this PR would be to have dynamo special-case subclass `__new__` and `__init__`, so we can fully + correctly inline them, all the way to their call to `_make_wrapper_subclass` call.

I think there are still some details of (2) that I haven't fully internalized / aren't clear to me (we still need the torch graph to contain IR that generates the subclass so AOTAutograd can trace it), but if we can get it to work then one nice benefit is that we wouldn't have to close over any custom python classes, which would be nice to not have to worry about for @jamesjwu 's AOTAutograd warm-cache work.

(counter-argument - even if we go with the approach in this PR, we can probably still figure out a way to hash the torch graph for the warm-start AOTAutograd work, by e.g. putting "non-FX-allowable" constructor arguments in a side-table somewhere that we need to hash).

bdhirsh commented:
> What is the difference between "can proxy in graph" vs "has an implementation of python_constant"?

I was trying to represent the first to literally mean "things that are allowed to show up as arguments to an FX node".

So I guess `torch.add(x, 1)` would showcase an example where they are different: even though `1` is technically a python constant, it is also a valid input to an FX node. And we in theory could give it the same treatment as stuff like `DeviceMesh` (close over it, and proxy a function like `dynamo_torch_add(x)` that closes over the `1`).

But we probably want to minimize the number of ways that we can end up with one of these "secretly close over arguments" proxied functions, since they make serialization a pain.

bdhirsh committed:
Update on "Dynamo: support proxying tensor subclass constructors, including with non-fx types"

For a while, we haven't really had good support for tensor subclass constructors going **inside** of the compiled region.

Two main example of that we've sort of half-handed in the last few months are:

(1) DTensor.to_local(), which is a wrapper around the constructor, that we explicitly allow_in_graph (cc wanchaol)
(2) Float8Tensor.to_float8(), which is similar, and also allowed_in_graph explicitly (cc drisspg / vkuzo)

This PR attempts to allow **all** traceable wrapper subclass constructors to go directly in the graph, when dynamo has a chance to intercept `SubclassCtr(*args)`.

To handle the DTensor case (where `Placement` and `DeviceMesh` cannot become nodes in the graph, but they are constructor inputs), I tried to add some infra on `VariableTracker` that tells us if a VT "supports going in the FX graph". For any inputs to the constructor that are not FX types, I close over them in the node's implementation, and convert them to their python constants (so AOT can implement them later).

cc anijain2305, a review on the dynamo bits would be much appreciated.




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

ezyang commented:
We talked about this at composability sync! https://www.youtube.com/watch?v=9JoHZLSLb14

bdhirsh committed:
Update on "Dynamo: support proxying tensor subclass constructors, including with non-fx types"

For a while, we haven't really had good support for tensor subclass constructors going **inside** of the compiled region.

Two main example of that we've sort of half-handed in the last few months are:

(1) DTensor.to_local(), which is a wrapper around the constructor, that we explicitly allow_in_graph (cc wanchaol)
(2) Float8Tensor.to_float8(), which is similar, and also allowed_in_graph explicitly (cc drisspg / vkuzo)

This PR attempts to allow **all** traceable wrapper subclass constructors to go directly in the graph, when dynamo has a chance to intercept `SubclassCtr(*args)`.

To handle the DTensor case (where `Placement` and `DeviceMesh` cannot become nodes in the graph, but they are constructor inputs), I tried to add some infra on `VariableTracker` that tells us if a VT "supports going in the FX graph". For any inputs to the constructor that are not FX types, I close over them in the node's implementation, and convert them to their python constants (so AOT can implement them later).

cc anijain2305, a review on the dynamo bits would be much appreciated.




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "Dynamo: support proxying tensor subclass constructors, including with non-fx types"

For a while, we haven't really had good support for tensor subclass constructors going **inside** of the compiled region.

Two main example of that we've sort of half-handed in the last few months are:

(1) DTensor.to_local(), which is a wrapper around the constructor, that we explicitly allow_in_graph (cc wanchaol)
(2) Float8Tensor.to_float8(), which is similar, and also allowed_in_graph explicitly (cc drisspg / vkuzo)

This PR attempts to allow **all** traceable wrapper subclass constructors to go directly in the graph, when dynamo has a chance to intercept `SubclassCtr(*args)`.

To handle the DTensor case (where `Placement` and `DeviceMesh` cannot become nodes in the graph, but they are constructor inputs), I tried to add some infra on `VariableTracker` that tells us if a VT "supports going in the FX graph". For any inputs to the constructor that are not FX types, I close over them in the node's implementation, and convert them to their python constants (so AOT can implement them later).

cc anijain2305, a review on the dynamo bits would be much appreciated.




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]
